sim/firesim-lib/src/main/scala/bridges/DromajoBridge.scala

```diff
@@ -0,0 +1,178 @@
+//See LICENSE for license details
+package firesim.bridges
+
+import chisel3._
+import chisel3.util._
+import chisel3.util.experimental.BoringUtils
+import freechips.rocketchip.config.{Parameters, Field}
+import freechips.rocketchip.diplomacy.AddressSet
+import freechips.rocketchip.util._
+import freechips.rocketchip.rocket.TracedInstruction
+import freechips.rocketchip.subsystem.RocketTilesKey
+import freechips.rocketchip.tile.TileKey
+
+import testchipip.{TileTraceIO, DeclockedTracedInstruction, TracedInstructionWidths}
+
+import midas.widgets._
+import testchipip.{StreamIO, StreamChannel}
+import junctions.{NastiIO, NastiKey}
+import TokenQueueConsts._
+
+//******
+//* MISC
+//******
+
+case class DromajoKey(
+  insnWidths: TracedInstructionWidths, // Widths of variable length fields in each TI
+  vecSizes: Int // The number of insns in each vec (= max insns retired at that core)
+)
+
+//*************
+//* TARGET LAND
+//*************
+
+class DromajoTargetIO(insnWidths: TracedInstructionWidths, numInsns: Int) extends Bundle {
+    val trace = Input(new TileTraceIO(insnWidths, numInsns))
+}
+
+/**
+ * Blackbox that is connected to the host
+ */
+class DromajoBridge(insnWidths: TracedInstructionWidths, numInsns: Int) extends BlackBox
+    with Bridge[HostPortIO[DromajoTargetIO], DromajoBridgeModule]
+{
+  val io = IO(new DromajoTargetIO(insnWidths, numInsns))
+  val bridgeIO = HostPort(io)
+
+  // give the Dromajo key to the GG module
+  val constructorArg = Some(DromajoKey(insnWidths, numInsns))
+
+  // generate annotations to pass to GG
+  generateAnnotations()
+}
+
+/**
+ * Helper function to connect blackbox
+ */
+object DromajoBridge {
+  def apply(tracedInsns: TileTraceIO)(implicit p: Parameters): DromajoBridge = {
+    val b = Module(new DromajoBridge(tracedInsns.insnWidths, tracedInsns.numInsns))
+    b.io.trace := tracedInsns
+    b
+  }
+}
+
+//*************************************************
+//* GOLDEN GATE MODULE
+//* This lives in the host (still runs on the FPGA)
+//*************************************************
+
+class DromajoBridgeModule(key: DromajoKey)(implicit p: Parameters) extends BridgeModule[HostPortIO[DromajoTargetIO]]()(p)
+{
+  lazy val module = new BridgeModuleImp(this) with UnidirectionalDMAToHostCPU {
+    // CONSTANTS: DMA Parameters
+    lazy val QUEUE_DEPTH = 64
+    lazy val TOKEN_WIDTH = 512 // in b
+    lazy val toHostCPUQueueDepth = QUEUE_DEPTH
+    lazy val dmaSize = BigInt((TOKEN_WIDTH / 8) * QUEUE_DEPTH)
+    // CONSTANTS
+
+    // setup io
+    val io = IO(new WidgetIO)
+    val hPort = IO(HostPort(new DromajoTargetIO(key.insnWidths, key.vecSizes)))
+
+    // the target is ready to both send/receive data
+    val tFire = hPort.toHost.hValid && hPort.fromHost.hReady
+
+    // helper to get number to round up to nearest multiple
+    def roundUp(num: Int, mult: Int): Int = { (((num - 1) / mult) + 1) * mult }
```

Comment:
id prefer you use `math.ceil` or something, though this is correct

Reply:
Is there a library function for this? This isn't the same as `ceil` since it rounds up to the nearest multiple of `mult` aka `roundUp(5,8) = 8`, `roundUp(59,8) = 64`.

Reply:
You still need to multiply the result but the first part of the expression is the part you can replace.